python/ray/data/_internal/execution/interfaces/execution_options.py

```diff
@@ -50,6 +67,47 @@ def add(self, other: "ExecutionResources") -> "ExecutionResources":
             )
         return total
 
+    def subtract(self, other: "ExecutionResources") -> "ExecutionResources":
+        """Subtracts execution resources.
+
+        Returns:
+            A new ExecutionResource object with subtracted resources.
+        """
+        total = ExecutionResources()
+        if self.cpu is not None or other.cpu is not None:
+            total.cpu = (self.cpu or 0.0) - (other.cpu or 0.0)
+        if self.gpu is not None or other.gpu is not None:
+            total.gpu = (self.gpu or 0.0) - (other.gpu or 0.0)
+        if (
+            self.object_store_memory is not None
+            or other.object_store_memory is not None
+        ):
+            total.object_store_memory = (self.object_store_memory or 0.0) - (
+                other.object_store_memory or 0.0
+            )
+        return total
+
+    def max(self, other: "ExecutionResources") -> "ExecutionResources":
+        """Returns the maximum for each resource type."""
+        return ExecutionResources(
+            cpu=max(self.cpu or 0.0, other.cpu or 0.0),
```

Comment:
Nit: if both fields are `None`, do we expected the max to be zero? I'm wondering it it'd be more intuitive if the field would still be `None`

For example:
```python
>>> result = ExecutionResources().max(ExecutionResources())
>>> result.cpu
0.0  # You might expect this to still be `None`?
```

Reply:
this is super confusing and annoying. but I think defaulting to 0 makes sense here. 
The reason is that this `ExecutionResources` is used for 2 different purposes simultaneously:
1) representing "resource usage": values should default to 0 in this case. 
2) representing "resource limits": values should default to inf in this case. 

the `min` and `max` API should only be used for (1).  I think maybe we should think more on how to sanitize this API and introduce (potentially breaking) changes before 2.10

Reply:
I will use a follow-up PR to make these values non-nullable.